### DISCUSSION TITLE
How can I uses lazydev without lspconfig?

### DISCUSSION THREAD

**Discussion by Charlie-83** [folke/lazydev.nvim#28] [Q&A]
Docs just say

> If you don't use [nvim-lspconfig](https://github.com/neovim/nvim-lspconfig), then you can use require('lazydev').find_workspace(buf?) to check if the buffer is part of an existing workspace or its libraries.

~But I don't know what this is actually instructing me to do.~ @folke clarified this is only for custom workspaces which I currently don't need

I use this autocommand to start `lua-ls`
```lua
autocmd("FileType", {
    pattern = "lua",
    callback = function()
        local root_dir = vim.fs.dirname(vim.fs.find({ "lazy-lock.json" })[1])
        local client = vim.lsp.start({
            name = "lua-ls",
            cmd = { "lua-language-server" },
            root_dir = root_dir,
        })
        vim.lsp.buf_attach_client(0, client)
    end,
})
```

I am using Lazy as my plugin manager and have
```lua
    {
        "folke/lazydev.nvim",
        ft = "lua",
        config = true,
    },
```

Currently, it isn't working and I'm getting warnings on all my `vim.api...` etc

**Comment by folke** [comment]
lazydev manages your workspaces, so you need to have a workspace.
If you don't want to use lspconfig, then that's of course fine, but then I'd expect you to know how to handle workspaces on your own.

**Reply by Charlie-83** [reply]
No idea what a workspace is (guess I haven't needed them so far). Thanks, now I know what I need to look into.

**Reply by Charlie-83** [reply]
Seems like `vim.lsp.start` sets the workspace folder automatically based on `root_dir`

```lua
print(vim.inspect(vim.lsp.buf.list_workspace_folders()))
```
Shows a table containing two directories (the cwd and the root directory for the Lua lsp).

I can add more workspace folders with `vim.lsp.buf.add_workspace_folder`. Do I need to add some special path here?

**Reply by Charlie-83** [reply]
Could never work this out. I ended up adding 
```lua
settings = {
    Lua = {
        workspace = {
            library = vim.api.nvim_get_runtime_file("", true)
        }
    }
}
```
to the `vim.lsp.start` call rather than using this plugin

**Reply by Jaehaks** [reply]
@Charlie-83 

`lazydev.nvim` accepts `lua_ls` as lua server name now, Please check your lsp name using `vim.lsp.config()`
for exmaple, line 13 in `lsp.lua` of `lazy.nvim` is like this 
```lua
---@param client? vim.lsp.Client
function M.supports(client)
  return client and vim.tbl_contains({ "lua_ls", "emmylua-analyzer-rust" }, client.name)
end
```

If you add the lua lsp name what you set, the `lazydev` works
I set the configuration name of lua as `lua-ls` because neovim cannot accept underscore(_) in lsp name and recommand `luals` as lsp name of lua. 
so I changed the chunk like this
```lua
---@param client? vim.lsp.Client
function M.supports(client)
  return client and vim.tbl_contains({ "lua-ls", "emmylua-analyzer-rust" }, client.name)
end
```

and `lazydev` works. 
Without empty `workspace` field of lua lsp configuraiton, `lazydev.nvim` inserts the `VIMRUNTIME` and necessary required libraries what your lua file contains.

**Reply by appleseedexm** [reply]
I can confirm the underlying issue described by @Jaehaks, and I wonder why there's not more information about this behaviour since this cant be the first time this happened. 
However, since neovim `0.12.0` (probably even before that) lsps with `_` (underscore) are supported. Renaming my `lua-language-server` to `lua_ls` fixed this. I wished there was some documentation for it though, or better, a property to set the language server name.

**Reply by yorgetsuten** [reply]
> I can confirm the underlying issue described by @Jaehaks, and I wonder why there's not more information about this behaviour since this cant be the first time this happened. 

Unfortunately, this is the situation with many plugins by @folke. They are so poorly documented that using them requires reading the source code to make something work.

**Reply by keelefi** [reply]
Thank you so much for this @appleseedexm ! I would never in a million years have figured out that the steps needed to fix my config was to rename the LSP from `luals` to `lua_ls`.

Especially since blogs like [lsp-zero.nvim](https://lsp-zero.netlify.app/blog/lsp-config-overview.html) suggest you should create the file `lsp/luals.lua` instead of `lsp/lua_ls.lua`.

**Comment by Miyelsh** [comment]
I was able to get lazydev installed without LazyVim or lspconfig with the following:

lazydev.nvim cloned into `~/.config/nvim/pack/default/start/`

Configuring the LSP in init.lua:
```
vim.api.nvim_create_autocmd("FileType", {
        pattern = "lua"
        , callback = function(args)
            vim.lsp.start({
                name = "lua-ls",
                cmd = {"lua-language-server"},
                root_dir = vim.fs.root(args.buf, "init.lua"),
                settings = {
                    Lua = {
                        workspace = {
                            library = vim.api.nvim_get_runtime_file("", true)
                        }
                    },
                }
            })
        end,
    }
)
```
